### ISSUE TITLE
Bug: count_ops and simplify fail for expression involving MatrixSymbol and indexing

### ISSUE THREAD

**Issue by truongduy134** [sympy/sympy#9324]
The function count_ops (in sympy/core/function.py) fails for expression involving MatrixSymbol and indexing (e.g. M[0,0] + 1). Following is the actual program that reproduces the bug:
```
>>> from sympy import *
>>> M = MatrixSymbol('M', 7, 7)
>>> count_ops(M[0,0] + 1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "sympy/core/function.py", line 2290, in count_ops
    if a.is_Rational:
AttributeError: 'str' object has no attribute 'is_Rational'
```

Cause: MatrixSymbol M has is_Symbol equal False. Hence, count_ops continues to examine M.args (as in the code block https://github.com/sympy/sympy/blob/7163137e22fbefbb645147a15147b26dad220d49/sympy/core/function.py#L2352-L2353 ). Also, M.args contains the string which is the name of the symbol, causing the failure when count_ops considers it in the next while-loop iteration (which is the string and has no is_Rational attribute)

Suggestion: Can we treat MatrixSymbol as symbol (with is_Symbol set to True) so that count_ops stops considering its arg?
I have created a pull request for my suggested solution. Look forward to discussion related to this issue

**Comment by truongduy134** [comment]
I have created a pull request for my suggested solution. Look forward to discussion related to this issue

**Comment by truongduy134** [comment]
* Update: Setting is_Symbol = True in MatrixSymbol is not a correct solution. It breaks some logic in other matrix functions (like inverse, ..)

**Comment by debugger22** [comment]
> Update: Setting is_Symbol = True in MatrixSymbol is not a correct solution. It breaks some logic in other matrix functions (like inverse, ..)

Tests seems to pass. Any example of the anomaly?

**Comment by truongduy134** [comment]
When setting is_Symbol = True, something is wrong with simplify function Here is the example from my actual project

```
>>> from sympy import *
>>> quaternion = MatrixSymbol('quaternion', 4, 1)
>>> rotationMatrix = Matrix([[1.0 - 2*(quaternion[1, 0]**2 + quaternion[2, 0]**2), 2*(quaternion[0, 0]*quaternion[1, 0] - quaternion[2, 0]*quaternion[3, 0]), 2*(quaternion[0, 0]*quaternion[2, 0] + quaternion[1, 0]*quaternion[3, 0])], [2*(quaternion[0, 0]*quaternion[1, 0] + quaternion[2, 0]*quaternion[3, 0]), 1.0 - 2*(quaternion[0, 0]**2 + quaternion[2, 0]**2), 2*(-quaternion[0, 0]*quaternion[3, 0] + quaternion[1, 0]*quaternion[2, 0])], [2*(quaternion[0, 0]*quaternion[2, 0] - quaternion[1, 0]*quaternion[3, 0]), 2*(quaternion[0, 0]*quaternion[3, 0] + quaternion[1, 0]*quaternion[2, 0]), 1.0 - 2*(quaternion[0, 0]**2 + quaternion[1, 0]**2)]])
>>> simplify(rotationMatrix)
Matrix([
[1.0,   0,   4],
[  4, 1.0,  -2],
[  0,   4, 1.0]])
```

Somehow, all the elements in symbol 'quaternion' are removed

**Comment by truongduy134** [comment]
OK. Here is the simpler example (when setting is_Symbol = True for MatrixSymbol):

```
>>> from sympy import *
>>> q = MatrixSymbol('q', 4, 1)
>>> x = Symbol('x')
>>> simplify(x + 1)
x + 1
>>> simplify(q[0, 0])
q[0, 0]
>>> simplify(q[0, 0] + 1)
1
```

I will trace the problem for simplify

**Comment by cbm755** [comment]
1. is_Symbol means literally is instance of Symbol class... (I think). And AFAIK MatrixSymbol is mot a subclass.

2. You don't need to close your PR if its work in progress, just put [WIP] at srart of title.

**Comment by truongduy134** [comment]
@cbm755 : Thanks for the advice. BTW, can we add new field to Basic : is_MatrixSymbol ?

**Comment by cbm755** [comment]
Probably we *can* if that turns out to be the best fix.  Sorry I don't have specific advice at the moment.

**Comment by truongduy134** [comment]
@cbm755 : I have opened the pull request https://github.com/sympy/sympy/pull/9392 for this issue. Would you like to have a look at it?

**Comment by toolforger** [comment]
Blacklisting tends to become unmaintainable over time; the issue is that code that makes such decision needs to know about all the thing that the object *cannot* be, but maintainers tend to know what the object at hand *can* be. It's better to whitelist things unless you can name a really convincing reason for blacklisting.
I.e. do not say `is_MatrixSymbol` and do `if not self.is_MatrixSymbol`. Instead, identify what quality/concept/feature/whatever it is that `MatrixSymbol`s do not have and that will cause these crashes (let me name it `Foo` here because I do not know what it is); define `is_Foo` and do `if is_foo:` henceforth.

**Comment by truongduy134** [comment]
Thanks @toolforger . I will try using existing is_Atom property instead.

**Comment by jcrist** [comment]
Closed by #9392.
